sdk/include/opentelemetry/sdk/trace/recordable.h

```diff
@@ -41,6 +42,18 @@ class Recordable
                       opentelemetry::trace::SpanId span_id,
                       opentelemetry::trace::SpanId parent_span_id) noexcept = 0;
 
+  /**
+   * Sets the resource associated with a span.
+   *
+   * <p> The resource is guaranteed to have a lifetime longer than this recordable.  It is
```

Comment:
> The resource is guaranteed to have a lifetime longer than this recordable.

I'm not sure if we can make this statement as is. I'm thinking about a scenario where the user frees the `TracerProvider`, but the exporter is still working on exporting spans (recordables), maybe in another thread.

If we go with plain pointers for resources (instead of `shared_ptr`), we'd need to be very careful, and either explicitly document all implicit assumptions we make, or take additional precautions to avoid problematic scenarios (e. g. making sure that in the `TracerProvider` destructor `Shutdown` is called for all exporters).

Reply:
I'm planning to change TracerProvider to (a) ensure shutdown and (b) require resource cleanup after shutdown.

Would you be amenable to that change in correlation with this?  I think it's needed to make this safe, and I'm happy to go update the docs.

Reply:
I'm definitely amenable to you making those changes.

Some other question: did you for performance reasons decide not to use a managed pointer (`std::shared_ptr`) for holding on to the resource in the recordable? And rather have implicit ownership via trace provider/recordable lifetimes?

So far we tried to make all ownership explicit to be on the (memory) safe side.